appmap/_implementation/testing_framework.py

```diff
@@ -90,11 +92,40 @@ def metadata(self):
         return ret
 
 
+NAME_MAX = 255  # true for most filesystems
+HASH_LEN = 7  # arbitrary, but git proves it's a reasonable value
+APPMAP_SUFFIX = '.appmap.json'
+
+
+def name_hash(namepart):
+    """Returns the hex digits of the sha256 of the os.fsencode()d namepart."""
+    return sha256(os.fsencode(namepart)).hexdigest()
+
+
+def write_appmap(basedir, basename, contents):
+    """Write an appmap file into basedir.
+
+    Adds APPMAP_SUFFIX to basename; shortens the name if necessary.
+    Atomically replaces existing files. Creates the basedir if required.
+    """
+
+    if len(basename) > NAME_MAX - len(APPMAP_SUFFIX):
+        part = NAME_MAX - len(APPMAP_SUFFIX) - 1 - HASH_LEN
+        basename = basename[:part] + '-' + name_hash(basename[part:])[:HASH_LEN]
+    filename = basename + APPMAP_SUFFIX
+
+    if not basedir.exists():
+        basedir.mkdir(parents=True, exist_ok=True)
+
+    with NamedTemporaryFile(mode='w', dir=basedir, delete=False) as tmp:
+        tmp.write(contents)
+    os.replace(tmp.name, basedir / filename)
```

Comment:
I wonder if we also need to worry about a failure here because the total path length exceeds the filesystem's limit? IIRC, we had a user report a failure when trying to write an appmap to a mounted filesystem, and it seemed like the root cause was the path length. I can't find the message now, though, so maybe I'm misremembering....

Reply:
I suppose this is an issue a user could run into, `PATH_MAX` is 4096 by default on Linux. But I don't think it makes sense for us to handle this; there's not much we can do, if a user has such a deep directory hierarchy then they're bound to run into problems sooner or later anyway. The error message in this case should be descriptive enough (I haven't checked though). And I think it's more important to have predictable names. Tl;dr I don't think should be in our purview.